src/client/components/draft/components/text_input_url.tsx

```diff
@@ -0,0 +1,190 @@
+}
+
+interface State {
+  url: string
+  selectionPosition: ClientRect
+}
+
+/**
+ * A popup UI for adding/removing links from text
+ * Can optionally be used for adding artist-follow links
+ */
+export class TextInputUrl extends Component<Props, State> {
+  constructor(props) {
+    super(props)
+
+    this.state = {
+      url: this.props.urlValue || "",
+      selectionPosition: getVisibleSelectionRect(window),
+    }
+  }
+
+  confirmLink = e => {
+    const { onConfirmLink, isFollowLink, onRemoveLink } = this.props
+    const url = e.target.value || this.state.url
+
+    e.preventDefault()
+    if (url.length) {
+      onConfirmLink(url, isFollowLink)
+    } else {
+      onRemoveLink()
+    }
+  }
+
+  onKeyDown = e => {
+    switch (e.key) {
+      case "Enter": {
+        return this.confirmLink(e)
+      }
+      case "Escape": {
+        return this.props.onClickOff()
+      }
+      case "Tab": {
+        return this.onExitInput(e)
+      }
+    }
+  }
+
+  onExitInput = e => {
+    const { onClickOff, onRemoveLink, urlValue } = this.props
+
+    if (e.target.value.length > 0) {
+      // Link was edited
+      this.confirmLink(e)
+    } else if (urlValue) {
+      // Link was deleted
+      onRemoveLink()
+    } else {
+      // No change, close the menu
+      onClickOff()
+    }
+  }
+
+  stickyControlsBox = () => {
+    const { editorPosition } = this.props
+    const { selectionPosition } = this.state
+    const textHeight = 17
+    const inputWidth = 350
+    let top
+    let left
+
+    if (editorPosition) {
+      top = selectionPosition.top - editorPosition.top + textHeight
+      left =
+        selectionPosition.left -
+        editorPosition.left +
+        selectionPosition.width / 2 -
+        inputWidth / 2
+    }
+    return { top, left }
+  }
+
+  render() {
+    const { url } = this.state
+    const { backgroundColor, onClickOff, onRemoveLink } = this.props
+    const { top, left } = this.stickyControlsBox()
+
+    return (
+      <>
+        <BackgroundOverlay onClick={onClickOff} />
+
+        <TextInputUrlContainer color={backgroundColor} top={top} left={left}>
+          <InputContainer>
+            <Input
+              autoFocus
+              className="bordered-input"
+              value={url}
+              onChange={e => this.setState({ url: e.target.value })}
+              placeholder="Paste or type a link"
+              onKeyDown={this.onKeyDown}
+            />
+
+            {url.length > 0 && (
+              <RemoveButton
+                onMouseDown={onRemoveLink}
+                background={color("black30")}
+              />
+            )}
+          </InputContainer>
+
+          <Button
+            onClick={this.confirmLink}
+            variant="secondaryGray"
+            size="small"
+          >
+            Apply
+          </Button>
+        </TextInputUrlContainer>
+      </>
+    )
+  }
+}
+
+const TextInputUrlContainer = styled.div.attrs<{ top: number; left: number }>(
```

Comment:
we can do this now
```typescript
styled.div<{ top: number; left: number }>``
````

https://github.com/styled-components/styled-components/issues/630#issuecomment-393638610

Reply:
Are we using this in Reaction already?  I'm getting a weird error, `ReferenceError: number is not defined`.  Tried updating `typescript-styled-plugin` and typescript versions are matched to Reaction.

Reply:
This requires a few toolchain updates, going to address in a separate PR